invokeai/backend/model_manager/metadata/metadata_base.py

```diff
@@ -0,0 +1,202 @@
+    """Raised when no metadata is available for a model."""
+
+
+class CommercialUsage(str, Enum):
+    """Type of commercial usage allowed."""
+
+    No = "None"
+    Image = "Image"
+    Rent = "Rent"
+    RentCivit = "RentCivit"
+    Sell = "Sell"
+
+
+class LicenseRestrictions(BaseModel):
+    """Broad categories of licensing restrictions."""
+
+    AllowNoCredit: bool = Field(
+        description="if true, model can be redistributed without crediting author", default=False
+    )
+    AllowDerivatives: bool = Field(description="if true, derivatives of this model can be redistributed", default=False)
+    AllowDifferentLicense: bool = Field(
+        description="if true, derivatives of this model be redistributed under a different license", default=False
+    )
+    AllowCommercialUse: CommercialUsage = Field(
+        description="Type of commercial use allowed or 'No' if no commercial use is allowed.", default_factory=set
+    )
+
+
+class RemoteModelFile(BaseModel):
+    """Information about a downloadable file that forms part of a model."""
+
+    url: AnyHttpUrl = Field(description="The url to download this model file", default=None)
+    path: Path = Field(description="The path to the file, relative to the model root")
+    size: int = Field(description="The size of this file, in bytes")
+    sha256: Optional[str] = Field(description="SHA256 hash of this model (not always available)", default=None)
+
+
+class ModelMetadataBase(BaseModel):
+    """Base class for model metadata information."""
+
+    name: str = Field(description="model's name")
+    author: str = Field(description="model's author")
+    tags: Set[str] = Field(description="tags provided by model source")
+
+
+class BaseMetadata(ModelMetadataBase):
+    """Adds typing data for discriminated union."""
+
+    type: Literal["basemetadata"] = "basemetadata"
+
+
+class ModelMetadataWithFiles(ModelMetadataBase):
+    """Base class for metadata that contains a list of downloadable model file(s)."""
+
+    files: List[RemoteModelFile] = Field(description="model files and their sizes", default_factory=list)
+
+    def download_urls(
+        self,
+        variant: Optional[ModelRepoVariant] = None,
+        subfolder: Optional[Path] = None,
+        session: Optional[Session] = None,
+    ) -> List[RemoteModelFile]:
+        """
+        Return a list of URLs needed to download the model.
+
+        :param variant: Return files needed to reconstruct the indicated variant (e.g. ModelRepoVariant('fp16'))
+        :param subfolder: Return files in the designated subfolder only
+        :param session: A request.Session object for offline testing
+
+        Note that the "variant" and "subfolder" concepts currently only apply to HuggingFace.
+        However Civitai does have fields for the precision and format of its models, and may
+        provide variant selection criteria in the future.
+        """
+        return self.files
+
+
+class CivitaiMetadata(ModelMetadataWithFiles):
```

Comment:
I noticed that the model hashes provided by civitai aren't included here. We should probably include those.

Maybe we should store the whole API response in its own column?

Reply:
The `sha256` model hash returned by Civitai are captured and stored in `RemoteModelFile` part of the metadata (see `metadata_base.py`). I considered storing the whole response, but it is huge and includes metadata on all the sample images that they display, along with the (often NSFW) prompts used to generate them. Civitai actually has multiple hashes generated using different algorithms. Is there an argument for storing more than one?

I hadn't been capturing and storing the HuggingFace sha256's and have just added them.

Reply:
I suppose any hash is as good as any other hash to identify the model, so there's no good reason to store all of them. Sha256 is the most common and you've got that covered. I must have missed that. 

Re: whole response, we can always add additional data later. Maybe we could eventually implement a sync metadata feature to get updated metadata (if the model itself hasn't changed). Or users could just re-import the model if we enhance the metadata storage.